sunpy/net/dataretriever/sources/adapt.py

```diff
@@ -0,0 +1,96 @@
+
+    @classmethod
+    def register_values(cls):
+        from sunpy.net import attrs as a
+
+        adict ={attrs.Instrument: [('ADAPT', 'ADvanced Adaptive Prediction Technique.')],
+                attrs.Source: [('NSO', 'National Solar Observatory.')],
+                attrs.Provider: [('GONG', 'Global Oscillation Network Group.')],
+                ADAPTFileType: [('4', 'Public')],
+                ADAPTLngType: [('0', 'Carrington Fixed'), ('1', 'Central Meridian'), ('2', 'East Limb')],
+                ADAPTInputSource: [('0', 'All'), ('1', 'KPVT'), ('2', 'VSM'), ('3', 'GONG'), ('4', 'HMI'), ('5', 'MDI'), ('6', 'MWO')],
+                ADAPTDataAssimilation: [('0', 'WH'), ('1', 'enLS'), ('2', 'enkf'), ('3', 'enLAKF')],
+                ADAPTResolution: [('1', '1.0 deg'), ('2', '0.2 deg')],
+                ADAPTVersionYear: [(str(i), f"Code version year -> {2000 + i}") for i in range(1, 20)],
+                ADAPTRealizations: [(str(i), f"Number of Realizations -> {i}") for i in range(1, 20)],
+                ADAPTVersionMonth: [(chr(i+96), f"Code version month -> {i}") for i in range(1, 13)],
+                ADAPTEvolutionMode: [('a', 'Data assimilation step'), ('i', 'Intermediate step'), ('f', 'Forecast step')],
+                ADAPTHelioData: [('n', 'Not added or no data'), ('f', 'Far-side'), ('e', 'Emergence'), ('b', 'Both emergence & far-side')],
+                ADAPTMagData: [('0', 'Not added or no data'), ('1', 'Mag-los'), ('2', 'Mag-vector'), ('3', 'Mag- both los & vector'), ('4', 'Mag- polar avg obs'), ('5', 'Mag- los & polar'), ('6', 'Mag- vector & polar'), ('7', 'Mag- both los and vector & polar')]
+                }
+        return adict
+
+    @classmethod
+    def _can_handle_query(cls, *query):
+        required = {attrs.Instrument, attrs.Time}
+
+        optional = {ADAPTFileType, ADAPTLngType, ADAPTInputSource, ADAPTDataAssimilation,
+                    ADAPTResolution, ADAPTVersionYear, ADAPTVersionMonth, ADAPTEvolutionMode,
+                    ADAPTHelioData, ADAPTMagData}
+
+        all_attrs = {type(x) for x in query}
+        return required.issubset(all_attrs) and all_attrs.issubset(required.union(optional))
+
+def carrington_time(CR=2193, frames=1):
+    """Get the start date for the start of a carrington rotation,
+    and a duration that will retrieve a given number of frames
+
+    Returns:
+        _type_: _description_
+    """
+    date = sunpy.coordinates.sun.carrington_rotation_time(CR)
+    date_end = date + frames*(3*1.9999999 * u.hour)
+
+    # Format the Search Dates
+    tstring = r"%Y-%m-%dT%H:%M:%S"
+    get_date    =date.strftime(tstring)
+    get_date_end=date_end.strftime(tstring)
+    return get_date, get_date_end
```

Comment:
This I think should be moved elsewhere, maybe to sunpy.time? 

This also overlaps with https://docs.sunpy.org/en/stable/generated/api/sunpy.coordinates.sun.carrington_rotation_time.html

I need input from the others for the best solution.

Reply:
Yeah, this is just a super basic wrapper around that carrington_rotation_time function, I'm glad to put it wherever (I suspect using the strftime is probably not necessary either since it is a Quantity)

Reply:
There are people better qualified to answer this than me but...

There are some issues with the implementation here but I think we can look past that:

So I assume you have this here because you think it will be useful to search via Carrington rotation rather than via a Timerange. As this stands I don't see this adding anything other than what exists on the function you have wrapped.

What I mean is that other than working out how much the end time should differ from the start time to get n results, a user can already convert a 2 CR numbers to times and then insert those into the attrs. Indeed if the user wants n results they can just select the first n rows from the query response before calling Fetch.  I also don't think it fits with the other clients where we have a unique helper function in a dataretrieverclient file that has to be called independently before calling a query.

Now what would be cool....

If searching by CR is desired then we could add a CarringtonRotation attr and have the can handle function check for either a.Time or a.CR and then, I don't know exactly where, but either in a pre_search_hook on the generic client or via attrwalker add_converter then we could optionally take either a CR number or a a.Time range to carry out an ADAPT search.

Now what form that new attr should take, I don't know. Your current API is start + number of observations but start+end makes a lot more sense rather than having an attr be composed of 2 different typed values. Is there a particular reason for your choices?

This would be much more teachable via an example than I think preconverrting the CR into times manually.

I wouldn't mind that being a separate follow on PR, as it isn't necessary for the client to operate, but I am not super keen on this function as it stands. I will bow down to those with greater knowledge though (and if so then we need to work on the implementation).

Keen to hear others thoughts.